ComponentKit/Utilities/CKArrayControllerChangesetVerification.mm

```diff
@@ -0,0 +1,158 @@
+/*
+ *  Copyright (c) 2014-present, Facebook, Inc.
+ *  All rights reserved.
+ *
+ *  This source code is licensed under the BSD-style license found in the
+ *  LICENSE file in the root directory of this source tree. An additional grant
+ *  of patent rights can be found in the PATENTS file in the same directory.
+ *
+ */
+
+#import "CKArrayControllerChangesetVerification.h"
+
+static NSArray<NSNumber *> *sectionCountsForSections(NSArray<NSArray *> *sections)
+{
+  NSMutableArray *sectionCounts = [NSMutableArray new];
+  for (NSArray *section in sections) {
+    [sectionCounts addObject:@(section.count)];
+  }
+  return [sectionCounts copy];
+}
+
+BOOL CKIsValidChangesetForSections(CKArrayControllerInputChangeset changeset, NSArray<NSArray *> *sections)
+{
+  NSMutableArray<NSNumber *> *currentSectionCounts = [NSMutableArray arrayWithArray:sectionCountsForSections(sections)];
+  __block BOOL invalidChangeFound = NO;
+
+  // Updates
+  changeset.items.enumerateItems(^(NSInteger section, NSInteger index, id<NSObject> object, BOOL *stop) {
+    if (section >= [currentSectionCounts count] || index >= [currentSectionCounts[section] integerValue]) {
```

Comment:
I imagine section or index won't typically be less than zero. It may be prudent to check regardless.